---
title: "Verifying BESCA signatures with bulk RNA-seq datasets"
author: "Jitao David Zhang"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output:
  html_document:
    df_print: paged
    theme: spacelab
    mathjax: default
    code_folding: hide
    toc: true
    toc_depth: 3
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: false
  pdf_document:
    number_sections: yes
    toc: yes
    toc_depth: 3
editor_options: 
  chunk_output_type: inline
params:
  echo: yes
  relative: FALSE
---

```{r setup, include=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      fig.height=6, fig.width=6)
library(ribiosUtils)
library(ribiosIO)
library(ribiosPlot)
library(ribiosGSEA)
library(openxlsx)
library(tidyverse)
library(ggplot2)
library(httr)
library(BioQC)
library(ComplexHeatmap)
library(igraph)
theme_set(theme_light(base_size=13))
```

# Background

We want to verify signatures of BESCA with bulk gene expression profiling data.

# Executive summary

We took two approaches to address the problem. First, we compared BESCA signatures with tissue and cell-type signatures provided by the BioQC package. Second, we examined expression of BESCA signatures in the data collection of human protein atlas.

Both approaches led us to observe consistency between bulk and single-cell signatures, as well as intriguing links between tissues and cell types that may lead to new insights.

# Analysis

## Data read in

We first download the BESCA signatures from the project's repository and parse its content.

```{r bescaSignature}
bescaGmt <- "data/20210719-BESCA-CellNames_scseqCMs6_sigs.gmt"
if(!file.exists(bescaGmt)) {
  download.file("https://raw.githubusercontent.com/bedapub/besca/master/besca/datasets/genesets/CellNames_scseqCMs6_sigs.gmt", destfile=bescaGmt)
}
bescaSignature <- BioQC::readGmt(bescaGmt)
```
Next we parse the BioQC signatures.

```{r}
bioqcSig <- BioQC::readCurrentSignatures()
```

```{r bbOverlap, cache=TRUE}
bescaBioQCOverlap <- sapply(bioqcSig, function(bq) {
  sapply(bescaSignature, function(be) {
    overlapCoefficient(be$genes, bq$genes)
  })
})
```

```{r}
isBescaStrOverlap <- apply(bescaBioQCOverlap, 1, function(x) any(x>=0.5))
isBioqcStrOverlap <- apply(bescaBioQCOverlap, 2, function(x) any(x>=0.5))
bescaBioQCstrOverlap <- bescaBioQCOverlap[isBescaStrOverlap, isBioqcStrOverlap]
bbCasord <- cascadeOrder(bescaBioQCstrOverlap)
bbCasOverlap <- bescaBioQCstrOverlap[bbCasord,]
colnames(bbCasOverlap) <- prettySigNames(colnames(bbCasOverlap))
```

```{r signatureOverlap, fig.height=12, fig.width=16}
ComplexHeatmap::Heatmap(bbCasOverlap,
                        name="Overlapping coefficient", 
                        cluster_columns=FALSE, cluster_rows = FALSE,
                        col=circlize::colorRamp2(breaks=c(0, 0.5, 1),
                                 colors=blackyellow(3)),
                        column_title = "BioQC signatures derived from bulk expression profiling of tissues and cell types",
                        row_title = "BESCA signatures derived from single-cell studies",
                        column_title_side = "bottom", column_title_gp = gpar(fontsize=20),
                        row_title_side = "right", row_title_gp = gpar(fontsize=20),
                        row_names_gp = gpar(fontsize=10),
                        column_names_gp = gpar(fontsize=10))
```

```{r bbGraph, fig.width=12, fig.height=12}
source("2021-07-toribios.R")
bbCasTops <- applyTopOrIncAndNotExclFilter(bbCasOverlap, MARGIN=1, top=1,
                                              falseValue=0,
                                              incFunc=function(x) x==1,
                                              excFunc=function(x) x==0)
bbCasUsedTops <- removeColumns(bbCasTops, function(x) !any(x>0))
bbGraph <- buildBescaIncidenceGraph(bbCasUsedTops)
bbLayout <- layout_with_fr(bbGraph, niter=2000)
{
  plot(bbGraph, layout=bbLayout)
  legend('topleft',legend=c("BESCA", "BioQC"), pch=21, pt.bg=c("orange", "lightblue"), cex=1.2, pt.cex=2.5)
}
```

The graph represents (1) gene expression signatures derived from single-cell sequencing studies extracted and offered by  BESCA (orange nodes), (2) gene expression signatures derived from bulk gene-expression profiling studies performed with microarray and next-generation sequencing extracted and offered by BioQC (blue nodes), and (3) their similarities encoded in edges. For each single-cell signature in orange node, we identify the bulk signature with the highest overlapping coefficient and connect them with an edge. In case more than one bulk signatures have the maximum overlapping coefficient of 1.0, all of them are shown and connected to the respective single-cell signature. 

We manually inspected the signatures. Despite different vocabularies were used in bulk and single-cell studies, most cell/tissue types match by corresponding biological entities. Examples include dendritic cell myeloid (bulk) and cDC/cDC1/cDC2 (single-cell), trachea (bulk) and cilated epithelial (single-cell), and liver and isolated hepatocytes (bulk) and hepatocytes (single-cell). They suggest that the signatures provided by BESCA and BioQC are well consistent with each other.

At the same time, there are a few intriguing mismatches, for instance multipotent cord blood unrestricted somatic stem cells (bulk) and fibroblasts (single-cell), and pituitary (bulk) and enteroendocrine and ionocyte (single-cell). There are several biologically plausible explanations. For instance, fibroblasts show heterogeneous, context-dependent expression patterns (Buechler, Matthew B., Rachana N. Pradhan, Akshay T. Krishnamurty, Christian Cox, Aslihan Karabacak Calviello, Amber W. Wang, Yeqing Angela Yang, et al. 2021. “Cross-Tissue Organization of the Fibroblast Lineage.” Nature, May, 1–5. https://doi.org/10.1038/s41586-021-03549-5.). Enteroendocrine cells are sensory cells of the gut that communicate by releasing hormones locally, which fulfils a similar endocrine function of the pituitary gland. Two genes, `CHGA` and `ASCL1`, are shared between bulk pituitary signature and single-cell ionocyte signatures and both are preferentially rexpressed by neuroendocrine cells. We failed to identify studies reporting whether they have identical or distinct biological functions in these contexts. It suggests that integrative analysis of single-cell and bulk gene expression signatures using *BESCA* and *BioQC* reveals hidden patterns that warrant further research.

The code below shows intersection between single-cell signatures of ionocytes and bulk signatures of pituitary gland.

```{r ionocyte_pituitary_gland_intersect}
intersect(bescaSignature[["Ionocyte"]]$genes,
          bioqcSig[["pituitary_gland_Fantom5_Tissue_0.7_3"]]$genes)
```

## Expression of single-cell signatures in human protein atlas.

```{r}
bescaSigHPAfile <- "data/bescaSigHPA.tsv.gz"
if(!file.exists(bescaSigHPAfile)) {
  consensus <- read_tsv("data/rna_consensus.tsv", col_types = "cccn") %>%
    mutate(Subset="consensus", GeneSymbol=`Gene name`, TC=Tissue)
  blood <- read_tsv("data/rna_blood_cell.tsv", col_types="cccnnn") %>%
    mutate(Subset="blood", GeneSymbol=`Gene name`, TC=`Blood cell`)
  hpasc <- read_tsv("data/rna_single_cell_type.tsv", col_types="cccn") %>%
    mutate(Subset="single-cell", GeneSymbol=`Gene name`, TC=`Cell type`)
  
  bescaSigGenes <- munion(gsGenes(bescaSignature))
  selTbl <- function(x) x %>% filter(GeneSymbol %in% bescaSigGenes) %>% dplyr::select(Subset, GeneSymbol, TC, NX)
  bescaSigHPA <- rbind(consensus %>% selTbl,
                       blood %>% selTbl,
                       hpasc %>% selTbl)
  write_tsv(bescaSigHPA, bescaSigHPAfile)
} else {
  bescaSigHPA <- read_tsv(bescaSigHPAfile, col_type="cccn")
}
bescaDf <- list2df(gsGenes(bescaSignature), col.names = c("GeneSet", "GeneSymbol"))
bescaSigMat <- inner_join(bescaSigHPA, bescaDf, by="GeneSymbol") %>%
  group_by(Subset, GeneSet, TC) %>%
  summarise(MedianNX=median(NX), .groups="drop") %>%
  longdf2matrix(., row.col="GeneSet", column.col="TC", value.col="MedianNX")
bescaSigMatOrd <- cascadeOrder(bescaSigMat)
bescaSigMatOrdered <- bescaSigMat[bescaSigMatOrd, ]
hpaSubset <- matchColumn(colnames(bescaSigMat), bescaSigHPA, "TC")$Subset
hpaSubsetCol <- RColorBrewer::brewer.pal(3, "Set1")
bescaSigMatAnno <- ComplexHeatmap::HeatmapAnnotation(Subset=factor(hpaSubset),
                                                     col = list(Subset=c("consensus"=hpaSubsetCol[1],
                                                                "blood"=hpaSubsetCol[2],
                                                                "single-cell"=hpaSubsetCol[3])),
                                                     which = "column",
                                                     annotation_name_side = "right")
```

```{r hpaPlot, fig.height=15, fig.width=18}
ComplexHeatmap::Heatmap(bescaSigMatOrdered,
                        top_annotation = bescaSigMatAnno,
                        name="Normalised expression by HPA", 
                        cluster_columns=FALSE, cluster_rows = FALSE,
                        col=circlize::colorRamp2(breaks=c(0, 
                                                          quantile(bescaSigHPA$NX, 0.9), 
                                                          quantile(bescaSigHPA$NX, 0.99)),
                                 colors=blackyellow(3)),
                        column_title = "Human Protein Atlas tissue and cell type",
                        row_title = "BESCA signatures",
                        column_title_side = "bottom", column_title_gp = gpar(fontsize=20),
                        row_title_side = "right", row_title_gp = gpar(fontsize=20),
                        row_names_gp = gpar(fontsize=10),
                        column_names_gp = gpar(fontsize=10))
```

```{r bhGraph, fig.width=15, fig.height=15}
hpaCol <- "#f4ebd0" ## cream
bhTops <- applyTopOrIncAndNotExclFilter(bescaSigMatOrdered, MARGIN=1, top=1,
                                        falseValue=0,
                                        excFunc=function(x) x==0)
bhUsedTops <- removeColumns(bhTops, function(x) !any(x>0))
bhGraph <- buildBescaIncidenceGraph(bhUsedTops,
                                    column_node_color = hpaCol, 
                                    column_label_color = "#122620") ## charcoal
bhLayout <- layout_with_fr(bhGraph, niter=2000)
{
  plot(bhGraph, layout=bhLayout)
  legend('topleft',legend=c("BESCA", "HPA"), pch=21, pt.bg=c("orange", hpaCol), cex=1.5, pt.cex=3)
}
```

## Exporting results

```{r}
write_gct(bbCasOverlap, "data/Besca-BioQC-signature-overlapping-coefficients.gct")
write_gct(bbCasOverlap, "data/Besca-signature-median-NX-in-HPA.gct")
igraph::write_graph(bbGraph, "data/Besca-BioQC-bbGraph.graphml", format="graphml")
igraph::write_graph(bhGraph, "data/Besca-HPA-bhGraph.graphml", format="graphml")
```

# Acknowledgment

I think colleagues of the BEDA team for continuous support and dicsussions.

# Session information

```{r bedaInfo}
bedaInfo()
```

```{r sessionInfo}
sessionInfo()
```

# Appendix
